### PR TITLE
TO loop

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Bugfix

### Description

It was occasionally observed that the TO could get stuck performing reconciliations forever when topics were created/modified simultaneously with TO startup. The logging was insufficient to prove conclusively what the problem was, but I realised that the timed reconciliation code wasn't performing all of its work with the per-topic lock held, which could certainly provide for an initial race between the initial reconciliation and watch-based reconciliation.

The commits in this PR do three things:

1. Improves the logging
2. Changes the old home-brew mutual exclusion mechanism (`InFlight`) to using Vertx locks, removing some rather intricate code that was only lightly tested in favour of something more easily reasoned about.
3. Changes the scope of the locks used for timed reconciliation to include the "reads" as well as the writes. This should mean that the timed reconciliation code will run exclusively from the watch-based reconciliations.

### Checklist

- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally

### CODE REVIEW

**Comment by tombentley** [comment]
@strimzi-ci run tests

File: topic-operator/src/main/java/io/strimzi/operator/topic/Session.java [hunk 1]
```diff
@@ -102,16 +103,20 @@ public void handle(Long inflightTimerId) {
                     if (zkResult.failed()) {
                         LOGGER.warn("Error disconnecting from zookeeper: {}", String.valueOf(zkResult.cause()));
                     }
-                    LOGGER.debug("Closing AdminClient {}", adminClient);
-                    adminClient.close(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
-
-                    HttpServer healthServer = this.healthServer;
-                    if (healthServer != null) {
-                        healthServer.close();
+                    long timeoutMs = Math.max(1, deadline - System.currentTimeMillis());
```
Comment (ppatierno) [inline topic-operator/src/main/java/io/strimzi/operator/topic/Session.java]: why this adjustment? did you notice any strange `deadline - System.currentTimeMillis()` value?

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java [hunk 2]
```diff
@@ -344,11 +347,73 @@ public TopicOperator(Vertx vertx, Kafka kafka,
...
     }
 
+    /**
+     * Run the given {@code action} on the context thread,
+     * immediately if there are currently no other actions with the given {@code key},
+     * or when the other actions with the given {@code key} have completed.
+     * When the given {@code action} is complete it must complete its argument future,
+     * which will complete the returned future
+     */
+    public <T> Future<T> acquireLock(TopicName key, Handler<Future<T>> action) {
```
Comment (ppatierno) [inline topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java]: about the name, it doesn't just acquire a lock but even add the action to do in the queue. Maybe just an `enqueueAction`  ? From the code it's clear it acquire a lock for doing that maybe not need to reflect in the name?

File: topic-operator/src/main/java/io/strimzi/operator/topic/Session.java (hunk 1 above)
Comment (tombentley) [reply topic-operator/src/main/java/io/strimzi/operator/topic/Session.java]: I observed an exception caused by a negative timeout, and I know from experience that a 0 argument also causes an exception.

File: topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java (hunk 2 above)
Comment (tombentley) [reply topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperator.java]: I agree `acquireLock` is a terrible name, but I couldn't think of a better one. How about `executeWithTopicLockHeld`, which does exactly what it says, even though it's a bit long. Wdyt?

**Comment by tombentley** [comment]
Testing was mostly via the `TopicOperatorIT`, which uses a real Kafka cluster (thanks to the Debezium test API) and a real Kubernetes cluster for the CR, but the TO itself is not running inside a container.

**Comment by tombentley** [comment]
Note: The Jenkins failures are unrelated to this PR.

**Pull request merged**